File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/comment/comment/direct/Java/Python/s749703010_commented.py
import sys

def main():
    sys.setrecursionlimit(1 << 25)
    n = int(sys.stdin.readline())
    a0 = list(map(int, sys.stdin.readline().split()))
    a1 = list(map(int, sys.stdin.readline().split()))
    a = [a0, a1]
    
    memo = [[[ -1 for _ in range(1 << n)] for _ in range(n + 1)] for _ in range(2)]
    
    def dp(flip, idx, msk):
        if msk == (1 << n) - 1:
            return 0
        if memo[flip][idx][msk] != -1:
            return memo[flip][idx][msk]
        x = bin(msk).count('1')
        ans = float('inf')
        ff = 0
        for i in range(n):
            if not (msk & (1 << i)):
                nf = x % 2
                if idx == n or (a[flip][idx] <= a[ ( (nf + (ff%2)) %2 ) ][i]):
                    new_flip = (nf + (ff % 2)) % 2
                    res = dp(new_flip, i, msk | (1 << i))
                    temp = ff + res
                    if temp < ans:
                        ans = temp
                ff += 1
        memo[flip][idx][msk] = ans
        return ans
    
    ans = dp(0, n, 0)
    print(-1 if ans >= 1e9 else ans)

if __name__ == "__main__":
    main()
# End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/comment/comment/direct/Java/Go/s749703010_commented.go
package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"strconv"
	"strings"
)

var a [][2]int
var memo [][[][]int]

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanLines)

	sc.Scan() // Read n
	n, _ := strconv.Atoi(sc.Text())

	a = make([][2]int, n)

	sc.Scan() // Read first array line
	firstLine := sc.Text()
	fields := strings.Fields(firstLine)
	for i := 0; i < n; i++ {
		a[i][0], _ = strconv.Atoi(fields[i])
	}

	sc.Scan() // Read second array line
	secondLine := sc.Text()
	fields = strings.Fields(secondLine)
	for i := 0; i < n; i++ {
		a[i][1], _ = strconv.Atoi(fields[i])
	}

	// Initialize memo
	memo = make([][][]int, 2)
	for f := 0; f < 2; f++ {
		memo[f] = make([][]int, n+1)
		for i := 0; i <= n; i++ {
			memo[f][i] = make([]int, 1<<n)
			for m := range memo[f][i] {
				memo[f][i][m] = -1
			}
		}
	}

	ans := dp(0, n, 0)
	if ans >= 1000000000 {
		fmt.Println(-1)
	} else {
		fmt.Println(ans)
	}
}

func dp(flip, idx, msk int) int {
	if msk == (1<<n)-1 {
		return 0
	}
	if memo[flip][idx][msk] != -1 {
		return memo[flip][idx][msk]
	}
	x := bits.OnesCount(uint(msk))
	ff := 0
	ans := 1000000000
	for i := 0; i < n; i++ {
		if (msk&(1<<i)) == 0 {
			nf := x % 2
			if idx == n || a[idx][flip] <= a[i][(nf+(ff%2))%2] {
				newFlip := (nf + (ff % 2)) % 2
				res := ff + dp(newFlip, i, msk|(1<<i))
				if res < ans {
					ans = res
				}
			}
			ff++
		} else {
			x--
		}
	}
	memo[flip][idx][msk] = ans
	return ans
}

// End of Code
